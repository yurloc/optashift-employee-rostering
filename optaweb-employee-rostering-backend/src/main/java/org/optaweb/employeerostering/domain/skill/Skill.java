/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.optaweb.employeerostering.domain.skill;

import javax.persistence.Entity;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import javax.validation.constraints.Size;

import org.optaweb.employeerostering.domain.common.AbstractPersistable;

@Entity
@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"tenantId", "name"}))
public class Skill extends AbstractPersistable {

    @NotNull
    @Size(min = 1, max = 120)
    @Pattern(regexp = "^(?!\\s).*(?<!\\s)$", message = "Name should not contain any leading or trailing whitespaces")
    private String name;

    @SuppressWarnings("unused")
    public Skill() {

    }

    public Skill(Integer tenantId, String name) {
        super(tenantId);
        this.name = name;
    }

    @Override
    public String toString() {
        return name;
    }

    // ************************************************************************
    // Simple getters and setters
    // ************************************************************************

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}
