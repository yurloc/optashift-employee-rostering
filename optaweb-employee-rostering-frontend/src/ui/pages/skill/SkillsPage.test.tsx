/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Skill } from 'domain/Skill';
import { Sorter, ReadonlyPartial } from 'types';
import { useTranslation } from 'react-i18next';
import { getRouterProps } from 'util/BookmarkableTestUtils';
import { SkillsPage, Props } from './SkillsPage';

describe('Skills page', () => {
  it('should render correctly with no skills', () => {
    const skillsPage = shallow(<SkillsPage {...noSkills} />);
    expect(toJson(skillsPage)).toMatchSnapshot();
  });

  it('should render correctly with a few skills', () => {
    const skillsPage = shallow(<SkillsPage {...twoSkills} />);
    expect(toJson(skillsPage)).toMatchSnapshot();
  });

  it('should render the viewer correctly', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const skill = { name: 'Skill', tenantId: 0, id: 1, version: 0 };
    const viewer = shallow(skillsPage.renderViewer(skill));
    expect(toJson(viewer)).toMatchSnapshot();
  });

  it('should render the editor correctly', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const skill = { name: 'Skill', tenantId: 0, id: 1, version: 0 };
    const editor = shallow(skillsPage.renderEditor(skill));
    expect(toJson(editor)).toMatchSnapshot();
  });

  it('should update properties on change', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const setProperty = jest.fn();
    const editor = skillsPage.editDataRow({}, setProperty);
    const nameCol = shallow(editor[0]);
    nameCol.simulate('change', { currentTarget: { value: 'Test' } });
    expect(setProperty).toBeCalled();
    expect(setProperty).toBeCalledWith('name', 'Test');
  });

  it('should an empty object on getInitialStateForNewRow', () => {
    const skillsPage = new SkillsPage(twoSkills);
    expect(skillsPage.getInitialStateForNewRow()).toEqual({});
  });

  it('should call addSkill on addData', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const skill = { name: 'Skill', tenantId: 0 };
    skillsPage.addData(skill);
    expect(twoSkills.addSkill).toBeCalled();
    expect(twoSkills.addSkill).toBeCalledWith(skill);
  });

  it('should call updateSkill on updateData', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const skill = { name: 'Skill', tenantId: 0, id: 1, version: 0 };
    skillsPage.updateData(skill);
    expect(twoSkills.updateSkill).toBeCalled();
    expect(twoSkills.updateSkill).toBeCalledWith(skill);
  });

  it('should call removeSkill on removeData', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const skill = { name: 'Skill', tenantId: 0, id: 1, version: 0 };
    skillsPage.removeData(skill);
    expect(twoSkills.removeSkill).toBeCalled();
    expect(twoSkills.removeSkill).toBeCalledWith(skill);
  });

  it('should return a filter that match by name', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const filter = skillsPage.getFilter();

    expect(twoSkills.tableData.filter(filter('1'))).toEqual([twoSkills.tableData[0]]);
    expect(twoSkills.tableData.filter(filter('2'))).toEqual([twoSkills.tableData[1]]);
  });

  it('should return a sorter that sort by name', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const sorter = skillsPage.getSorters()[0] as Sorter<Skill>;
    const list = [twoSkills.tableData[1], twoSkills.tableData[0]];
    expect(list.sort(sorter)).toEqual(twoSkills.tableData);
  });

  it('should treat incompleted data as incomplete', () => {
    const skillsPage = new SkillsPage(twoSkills);

    const noName: ReadonlyPartial<Skill> = { tenantId: 0 };
    const result1 = skillsPage.isDataComplete(noName);
    expect(result1).toEqual(false);

    const completed: ReadonlyPartial<Skill> = { tenantId: 0, name: 'Name' };
    const result2 = skillsPage.isDataComplete(completed);
    expect(result2).toEqual(true);
  });

  it('should treat empty name as invalid', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const components: Skill = { tenantId: 0, name: '' };
    const result = skillsPage.isValid(components);
    expect(result).toEqual(false);
  });

  it('should treat non-empty name as valid', () => {
    const skillsPage = new SkillsPage(twoSkills);
    const components: Skill = { tenantId: 0, name: 'Skill' };
    const result = skillsPage.isValid(components);
    expect(result).toEqual(true);
  });
});

const noSkills: Props = {
  ...useTranslation(),
  tReady: true,
  tenantId: 0,
  title: 'Skills',
  columnTitles: ['Name'],
  tableData: [],
  addSkill: jest.fn(),
  updateSkill: jest.fn(),
  removeSkill: jest.fn(),
  ...getRouterProps('/skills', {}),
};

const twoSkills: Props = {
  ...useTranslation(),
  tReady: true,
  tenantId: 0,
  title: 'Skills',
  columnTitles: ['Name'],
  tableData: [{
    id: 0,
    version: 0,
    tenantId: 0,
    name: 'Skill 1',
  },
  {
    id: 1,
    version: 0,
    tenantId: 0,
    name: 'Skill 2',
  }],
  addSkill: jest.fn(),
  updateSkill: jest.fn(),
  removeSkill: jest.fn(),
  ...getRouterProps('/skills', {}),
};
