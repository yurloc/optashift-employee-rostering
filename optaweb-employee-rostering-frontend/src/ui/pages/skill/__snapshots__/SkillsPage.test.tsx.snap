// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skills page should render correctly with a few skills 1`] = `
<Fragment>
  <Level
    gutter="sm"
    style={
      Object {
        "backgroundColor": "var(--pf-global--BackgroundColor--200)",
        "padding": "5px 5px 5px 5px",
      }
    }
  >
    <LevelItem>
      <FilterComponent
        filterText=""
        onChange={[Function]}
      />
    </LevelItem>
    <LevelItem
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <Button
        isDisabled={false}
        onClick={[Function]}
      >
        Trans(i18nKey=add)
      </Button>
      <Pagination
        itemCount={2}
        onPerPageSelect={[Function]}
        onSetPage={[Function]}
        page={1}
        perPage={10}
        widgetId="pagination-options-menu-top"
      />
    </LevelItem>
  </Level>
  <Table
    borders={true}
    caption="Skills"
    cells={
      Array [
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Name",
          "transforms": Array [
            [Function],
          ],
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "",
          "transforms": undefined,
        },
      ]
    }
    className=""
    contentId="expanded-content"
    dropdownDirection="down"
    dropdownPosition="right"
    expandId="expandable-toggle"
    gridBreakPoint="grid-md"
    onSort={[Function]}
    role="grid"
    rowLabeledBy="simple-node"
    rows={
      Array [
        Object {
          "cells": Array [
            Object {
              "title": <EditableComponent
                editor={
                  <TextInput
                    aria-label="Name"
                    className=""
                    defaultValue="Skill 1"
                    isDisabled={false}
                    isReadOnly={false}
                    isRequired={false}
                    isValid={true}
                    name="name"
                    onChange={[Function]}
                    type="text"
                  />
                }
                isEditing={false}
                viewer={
                  <Text>
                    Skill 1
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <span
                    style={
                      Object {
                        "display": "grid",
                        "gridColumnGap": "5px",
                        "gridTemplateColumns": "1fr auto auto",
                      }
                    }
                  >
                    <span />
                    <Button
                      aria-label="Save"
                      onClick={[Function]}
                      variant="link"
                    >
                      <SaveIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Cancel"
                      onClick={[Function]}
                      variant="link"
                    >
                      <CloseIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
                isEditing={false}
                viewer={
                  <span
                    style={
                      Object {
                        "display": "grid",
                        "gridColumnGap": "5px",
                        "gridTemplateColumns": "1fr auto auto",
                      }
                    }
                  >
                    <span />
                    <Button
                      aria-label="Edit"
                      onClick={[Function]}
                      variant="link"
                    >
                      <EditIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Delete"
                      onClick={[Function]}
                      variant="link"
                    >
                      <TrashIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
              />,
            },
          ],
        },
        Object {
          "cells": Array [
            Object {
              "title": <EditableComponent
                editor={
                  <TextInput
                    aria-label="Name"
                    className=""
                    defaultValue="Skill 2"
                    isDisabled={false}
                    isReadOnly={false}
                    isRequired={false}
                    isValid={true}
                    name="name"
                    onChange={[Function]}
                    type="text"
                  />
                }
                isEditing={false}
                viewer={
                  <Text>
                    Skill 2
                  </Text>
                }
              />,
            },
            Object {
              "title": <EditableComponent
                editor={
                  <span
                    style={
                      Object {
                        "display": "grid",
                        "gridColumnGap": "5px",
                        "gridTemplateColumns": "1fr auto auto",
                      }
                    }
                  >
                    <span />
                    <Button
                      aria-label="Save"
                      onClick={[Function]}
                      variant="link"
                    >
                      <SaveIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Cancel"
                      onClick={[Function]}
                      variant="link"
                    >
                      <CloseIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
                isEditing={false}
                viewer={
                  <span
                    style={
                      Object {
                        "display": "grid",
                        "gridColumnGap": "5px",
                        "gridTemplateColumns": "1fr auto auto",
                      }
                    }
                  >
                    <span />
                    <Button
                      aria-label="Edit"
                      onClick={[Function]}
                      variant="link"
                    >
                      <EditIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                    <Button
                      aria-label="Delete"
                      onClick={[Function]}
                      variant="link"
                    >
                      <TrashIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="sm"
                        title={null}
                      />
                    </Button>
                  </span>
                }
              />,
            },
          ],
        },
      ]
    }
    sortBy={Object {}}
    variant={null}
  >
    <TableHeader />
    <TableBody />
  </Table>
</Fragment>
`;

exports[`Skills page should render correctly with no skills 1`] = `
<Fragment>
  <Level
    gutter="sm"
    style={
      Object {
        "backgroundColor": "var(--pf-global--BackgroundColor--200)",
        "padding": "5px 5px 5px 5px",
      }
    }
  >
    <LevelItem>
      <FilterComponent
        filterText=""
        onChange={[Function]}
      />
    </LevelItem>
    <LevelItem
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <Button
        isDisabled={false}
        onClick={[Function]}
      >
        Trans(i18nKey=add)
      </Button>
      <Pagination
        itemCount={0}
        onPerPageSelect={[Function]}
        onSetPage={[Function]}
        page={1}
        perPage={10}
        widgetId="pagination-options-menu-top"
      />
    </LevelItem>
  </Level>
  <Table
    borders={true}
    caption="Skills"
    cells={
      Array [
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Name",
          "transforms": Array [
            [Function],
          ],
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "",
          "transforms": undefined,
        },
      ]
    }
    className=""
    contentId="expanded-content"
    dropdownDirection="down"
    dropdownPosition="right"
    expandId="expandable-toggle"
    gridBreakPoint="grid-md"
    onSort={[Function]}
    role="grid"
    rowLabeledBy="simple-node"
    rows={Array []}
    sortBy={Object {}}
    variant={null}
  >
    <TableHeader />
    <TableBody />
  </Table>
</Fragment>
`;

exports[`Skills page should render the editor correctly 1`] = `
<tr>
  <td
    key="0"
  >
    <TextInput
      aria-label="Name"
      className=""
      defaultValue="Skill"
      isDisabled={false}
      isReadOnly={false}
      isRequired={false}
      isValid={true}
      key="0"
      name="name"
      onChange={[Function]}
      type="text"
    />
  </td>
</tr>
`;

exports[`Skills page should render the viewer correctly 1`] = `
<tr>
  <td
    key="0"
  >
    <Text
      key="0"
    >
      Skill
    </Text>
  </td>
</tr>
`;
